README.md
[![Build Status](https://travis-ci.org/hrbrmstr/dtupdate.png)](https://travis-ci.org/hrbrmstr/dtupdate)

The `dtupdate` packages has functions that attempt to figure out which packages have non-CRAN versions (currently only looks for github ones) and then tries to figure out which ones have updates (i.e. the github version is \> local version). It provides an option (not recommended) to auto-update any packages with newer development versions, which is tempered by the ability to selectively install said packages.

The `URL` and `BugReports` fields are, frankly, a mess. Many packages have multiple URLs in one or both of those fields and the non-github URLs are all over the place in terms of formatting. It will take some time, but I'm pretty confident I can get r-forge, bitbucket, gitorius and other external repos working. This was an easy first step.

The following functions are implemented:

-   `github_update` - find, report and optionally update packages installed from or available on github. It keys off of the `BugReports:` or `URL:` fields, looking for a github-ish URL and then grabbing what info that it can to see if the repo is in package format and has a `DESCRIPTION` file it can work with

### News

-   Version `1.4` adds a `whats_new()` function that uses `rmarkdown` to produce an HTML report from all the available `NEWS[.md]` files for any packages that have updates
-   Version `1.3` works with both the `BugReports:` and `URL:` field and handles some additional URL-retrieving exceptions
-   Version `1.2` switches from `plyr` to `dplyr` to stop [@jayjacobs](<http://twitter.com/jayjacobs>) from whining, removes the `libLoc` switch from version `1.1` and replaces it with a boolean `all` parameter which let's you switch from inspecting the github packages installed across all library paths or just the first one (*usually* your user/local path), and adds a `show.location` parameter which will include the library path for each package in the data frame return
-   Version `1.1` incorporates functionality from [Thomas J Leeper](http://twitter.com/thosjleeper)'s gist: , including the ability to interactively select which packages to install, the option to specify which `libLoc` will be searched by `installed.packages` and handling of a fringe case where the repo name does not match the package name (ironically, this was due to Hadley's `reshape` repo for the `reshape2` package)
-   Version `1.0` released (nascent github pkg update capability)

### Installation

``` {.r}
devtools::install_github("hrbrmstr/dtupdate")
```

### Usage

``` {.r}
library(dtupdate)
```

    ## Loading required package: devtools
    ## 
    ## Attaching package: 'devtools'
    ## 
    ## The following objects are masked from 'package:utils':
    ## 
    ##     ?, help
    ## 
    ## The following object is masked from 'package:base':
    ## 
    ##     system.file
    ## 
    ## Loading required package: httr
    ## Loading required package: stringr
    ## Loading required package: dplyr
    ## 
    ## Attaching package: 'dplyr'
    ## 
    ## The following objects are masked from 'package:stats':
    ## 
    ##     filter, lag
    ## 
    ## The following objects are masked from 'package:base':
    ## 
    ##     intersect, setdiff, setequal, union
    ## 
    ## Loading required package: rmarkdown

``` {.r}
# current verison
packageVersion("dtupdate")
```

    ## [1] '1.4'

``` {.r}
# see what packages are available for an update
github_update()
```

    ## Source: local data frame [41 x 6]
    ## 
    ##           package           repo              owner installed.version current.version update.available
    ## 1           Hmisc          Hmisc           harrelfe            3.14-4          3.14-5             TRUE
    ## 2          RMySQL         RMySQL      jeffreyhorner             0.9-3           0.9-3            FALSE
    ## 3            Rcpp           Rcpp           RcppCore          0.11.2.1        0.11.2.2             TRUE
    ## 4   RcppArmadillo  RcppArmadillo           RcppCore         0.4.400.0     0.4.400.0.2             TRUE
    ## 5        Rttf2pt1       Rttf2pt1                wch             1.3.1           1.3.1            FALSE
    ## 6             WDI            WDI vincentarelbundock               2.4             2.4            FALSE
    ## 7     countrycode    countrycode vincentarelbundock              0.17            0.17            FALSE
    ## 8           dplyr          dplyr             hadley          0.2.0.99        0.2.0.99            FALSE
    ## 9        dtupdate       dtupdate           hrbrmstr               1.4             1.3            FALSE
    ## 10      extrafont      extrafont                wch              0.16       0.16.0.99             TRUE
    ## 11    extrafontdb    extrafontdb                wch               1.0             1.0            FALSE
    ## 12       forecast       forecast        robjhyndman               5.5             5.6             TRUE
    ## 13        formatR        formatR              yihui              0.10             1.0             TRUE
    ## 14       ggdendro       ggdendro             andrie            0.1-14          0.1-14            FALSE
    ## 15        ggplot2        ggplot2             hadley          1.0.0.99        1.0.0.99            FALSE
    ## 16         gmailr         gmailr          jimhester             0.0.1           0.0.1            FALSE
    ## 17          highr          highr              yihui               0.3           0.3.1             TRUE
    ## 18         httpuv         httpuv            rstudio             1.3.0           1.3.1             TRUE
    ## 19       jsonlite       jsonlite         jeroenooms            0.9.10       0.9.10.99             TRUE
    ## 20          knitr          knitr              yihui               1.6          1.6.15             TRUE
    ## 21 knitrBootstrap knitrBootstrap          jimhester             1.0.0           1.0.0            FALSE
    ## 22      lubridate      lubridate             hadley             1.3.3           1.3.3            FALSE
    ## 23       markdown       markdown            rstudio             0.7.4           0.7.4            FALSE
    ## 24        memoise        memoise             hadley            0.2.99          0.2.99            FALSE
    ## 25           mime           mime              yihui             0.1.2           0.1.2            FALSE
    ## 26       miniCRAN       miniCRAN             andrie            0.0-20          0.0-20            FALSE
    ## 27       miniCRAN       miniCRAN             andrie            0.0-14          0.0-20             TRUE
    ## 28        packrat        packrat            rstudio          0.4.0.12        0.4.0.12            FALSE
    ## 29       reshape2        reshape             hadley               1.4        1.4.0.99             TRUE
    ## 30         resolv         resolv           hrbrmstr             0.2.3           0.2.3            FALSE
    ## 31       roxygen2        roxygen         klutometis             4.0.1           4.0.2             TRUE
    ## 32       roxygen2        roxygen         klutometis             4.0.1           4.0.2             TRUE
    ## 33           rzmq           rzmq           armstrtw             0.7.0           0.7.0            FALSE
    ## 34         scales         scales             hadley          0.2.4.99        0.2.4.99            FALSE
    ## 35          shiny          shiny            rstudio       0.10.1.9004     0.10.1.9006             TRUE
    ## 36       shinyAce       shinyAce        trestletech             0.1.0           0.1.0            FALSE
    ## 37        slidify        slidify           ramnathv             0.4.5           0.4.5            FALSE
    ## 38      statebins      statebins           hrbrmstr               1.1             1.1            FALSE
    ## 39         testit         testit              yihui               0.3             0.3            FALSE
    ## 40          tidyr          tidyr             hadley               0.1      0.1.0.9000             TRUE
    ## 41        whisker        whisker            edwindj             0.3-2             0.4             TRUE

### Test Results

``` {.r}
library(dtupdate)
library(testthat)

date()
```

    ## [1] "Sun Aug 31 17:06:48 2014"

``` {.r}
test_dir("tests/")
```

    ## basic functionality : .


NEWS
dtupdate News

- Version 1.4 adds a whats_new() function that uses rmarkdown to produce an
  HTML report from all the available NEWS[.md] files for any packages
  that have updates

- Version 1.3 works with both the BugReports: and URL: field and
  handles some additional URL-retrieving exceptions

- Version 1.2 switches from plyr to dplyr to stop @jayjacobs from whining,
  removes the libLoc switch from version 1.1 and replaces it with a boolean
  all parameter which let's you switch from inspecting the github packages
  installed across all library paths or just the first one
  (*usually* your user/local path), and adds a show.location parameter
  which will include the library path for each package in the data frame return

- Version 1.1 incorporates functionality from Thomas J Leeper's gist,
  including the ability to interactively select which packages to install,
  the option to specify which `libLoc` will be searched by installed.packages
  and handling of a fringe case where the repo name does not match the package
  name (ironically, this was due to Hadley's reshape repo for
  the reshape2 package)

- Version 1.0 released (nascent github pkg update capability)

tests/test-all.R
library(testthat)
test_check("dtupdate")


tests/testthat/test-dtupdate.R
context("basic functionality")
test_that("we can run the function and get back a dat frame", {

  expect_that(github_update(), is_a("data.frame"))

})

R/dtupdate.R
# constructs (from a BugReports or URL URL) a raw.githubusercontent URL pointing to a DESCRIPTION file
# intended for use with dplyr's mutate() in github_update
build_description_url <- function(BugReports, URL) {

  buildit <- function(.BugReports, .URL) {

    urlout <- strsplit(.URL, 'github.com/')[[1]][2]
    urlout.b <- gsub('/issues','', .BugReports)
    urlout.b <- strsplit(urlout.b, 'github.com/')[[1]][2]

    urls <- c(str_match(urlout, "(^[a-zA-Z0-9\\_\\-]+/[a-zA-Z0-9\\_\\-]+$)")[,2],
              str_match(urlout.b, "(^[a-zA-Z0-9\\_\\-]+/[a-zA-Z0-9\\_\\-]+$)")[,2])

    urls <- unique(urls)
    urls <- urls[!is.na(urls)]

    return(ifelse(length(urls) == 0, NA, sprintf("https://raw.githubusercontent.com/%s/master/DESCRIPTION", urls)))

  }

  mapply(buildit, BugReports, URL, USE.NAMES=FALSE, SIMPLIFY=TRUE)

}

# extracts an owner from a raw.githubusercontent URL
# intended for use with dplyr's mutate() in github_update
extract_owner <- function(.description) {
  gsub("(https://raw.githubusercontent.com/|/.*$)", "", .description)
}

# extracts the github repo name from a raw.githubusercontnet URL
# intended for use with dplyr's mutate() in github_update
extract_repo <- function(.description) {
  return(str_match(.description, "https://raw.githubusercontent.com/[a-zA-Z0-9_\\.]+/([a-zA-Z0-9_\\.]+)")[,2])
}

# retrieves the version # from an R package DESCRIPTION file accessible via URLs in .description
# intended for use with dplyr's mutate() in github_update
retrieve_version <- function(.description, .progress=FALSE, pb=NULL) {
  unlist(sapply(.description, function(url) {
    if (.progress) setTxtProgressBar(pb, getTxtProgressBar(pb)+1)
    version <- character(0)
    try(req <- GET(url), silent=TRUE)
    try(version <- str_extract(grep("Version", readLines(textConnection(content(req, as="text"))), value=TRUE), "([0-9\\.\\-]+)"), silent=TRUE)
    ifelse(length(version) == 0 , NA, version)
  }, USE.NAMES=FALSE))
}

# returns a boolean vector of whether .github > .installed for each pair
# intended for use with dplyr's mutate() in github_update
compareVersions <- function(.installed, .github) {
  mapply(function(a, b) { compareVersion(a,b) < 0 }, .installed, .github, USE.NAMES=FALSE, SIMPLIFY=TRUE)
}

#' @title github_update - find, report and optionally update packages installed from or available on github
#' @description
#'     Finds local packages that [may] have github versions, then informs &
#'     optionally updates them to the newest versions. It is recommended that you
#'     still selectively perform the updates since you may want to double-check
#'     the repo for any gotchas or incompatibilities with other installed packages.
#'
#'     Added code from Thomas J Leeper's (@@thosjleeper) \code{#spiffy} gist: \url{https://gist.github.com/leeper/9123584}
#'
#' @param auto.install should it try to auto-install newer packages [not recommended] (bool, initially \code{FALSE})
#' @param ask if this and \code{auto.install}==\code{TRUE} a pkg install select list will be shown (graphical or text widget dependent upon \code{widget})
#' @param widget if this and \code{auto.install}==\code{TRUE} and \code{ask}==\code{TRUE} then a GUI select widget will
#' @param .progress show a progress bar (bool, initially \code{FALSE})
#' @param all search all \code{.libPaths()} (\code{TRUE}) or only the first one (initially \code{TRUE})
#' @param show.location include the library location in the returned data frame (initially \code{FALSE})
#' @return data frame of installed packages with information on owner & whether an update is available (optionally includes installed library path)
#' @import dplyr
#' @note package installation exceptions are caught with \code{try}, but are not reported out in any other way but console errors
#' @export
#' @examples \dontrun{
#' github_update()
#'
#' ##           package           repo       owner installed.version current.version update.available
#' ## 1            Rcpp           Rcpp    RcppCore            0.11.2        0.11.2.1             TRUE
#' ## 2   RcppArmadillo  RcppArmadillo    RcppCore         0.4.400.0       0.4.400.0            FALSE
#' ## 3      data.table     data.table  Rdatatable             1.9.3           1.9.3            FALSE
#' ## 4        dtupdate       dtupdate    hrbrmstr               1.1             1.1            FALSE
#' ## 5        forecast       forecast robjhyndman               5.4             5.6             TRUE
#' ## 6         formatR        formatR       yihui              0.10          0.10.5             TRUE
#' ## 7          gmailr         gmailr   jimhester             0.0.1           0.0.1            FALSE
#' ## 8        jsonlite       jsonlite  jeroenooms             0.9.9          0.9.10             TRUE
#' ## 9           knitr          knitr       yihui             1.6.6          1.6.14             TRUE
#' ## 10 knitrBootstrap knitrBootstrap   jimhester             1.0.0           1.0.0            FALSE
#' ## 11      lubridate      lubridate      hadley             1.3.3           1.3.3            FALSE
#' ## 12       markdown       markdown     rstudio             0.7.4           0.7.4            FALSE
#' ## 13        memoise        memoise      hadley             0.2.1          0.2.99             TRUE
#' ## 14           mime           mime       yihui             0.1.2           0.1.2            FALSE
#' ## 15       miniCRAN       miniCRAN      andrie            0.0-20          0.0-20            FALSE
#' ## 16        packrat        packrat     rstudio             0.4.0        0.4.0.12             TRUE
#' ## 17       reshape2        reshape      hadley               1.4        1.4.0.99             TRUE
#' ## 18         resolv         resolv    hrbrmstr             0.2.2           0.2.2            FALSE
#' ## 19          rzmq           rzmq    armstrtw             0.7.0           0.7.0            FALSE
#' ## 20         scales         scales      hadley             0.2.4        0.2.4.99             TRUE
#' ## 21          shiny          shiny     rstudio       0.10.0.9001     0.10.1.9004             TRUE
#' ## 22       shinyAce       shinyAce trestletech             0.1.0           0.1.0            FALSE
#' ## 23        slidify        slidify    ramnathv             0.4.5           0.4.5            FALSE
#' ## 24         testit         testit       yihui               0.3             0.3            FALSE
#' }
#'
github_update <- function(auto.install=FALSE, ask=TRUE, widget=FALSE, .progress=FALSE, all=TRUE, show.location=FALSE) {

  # without suppressWarnings, there are most likely going to be useless row.names warnings
  # prbly don't need "URL", but perhaps in the future

  locs <- .libPaths()
  if (!all) { locs <- locs[1] }

  suppressWarnings(pkgs <- data.frame(installed.packages(lib.loc=locs, fields=c("URL", "BugReports")), stringsAsFactors=FALSE))

  # so, we cheat and use the fact that the author/maintainer has a github "BugReports" DESCRIPTION string
  # as an indicator that there may be github devtools version of the package available

  # if there are alot of packages, you might appreciate knowing the progress

  if (.progress) pkg.pb <- txtProgressBar(min=0, max=nrow(pkgs), initial=0, title="Retrieving local and remote package information")

  # iterate over the possible github package list

  tmp <- as.tbl(pkgs) %>%
    filter(grepl("github", BugReports) | grepl("github", URL)) %>%           # we only care about github urls & assuming we only have one
    select(Package, Version, BugReports, URL, LibPath) %>%                   # we only need these fields for now
    mutate(description=build_description_url(BugReports, URL),               # build DESCRIPTION URL
           owner=extract_owner(description),                                 # extract repo owner
           repo=extract_repo(description),                                   # extract repo name
           current.version=retrieve_version(description, .progress, pkg.pb), # get repo version string
           update.available=compareVersions(Version, current.version)) %>%   # determine whether an update is available or not
    filter(!is.na(current.version)) %>%                                      # ignore any github repos that aren't in package format
    arrange(Package)                                                         # sort the result (since folks will be reading the output)

  # done with progress bars

  if (.progress) close(pkg.pb)

  # auto installation of out of date packages is not recommended, but for those living on the edge, here ya go!

  if (auto.install) {

    message("automatically updating packages...")

    # start with assuming all pkgs with updates avail
    k <- (tmp %>% filter(update.available==TRUE))$Package

    # filter if asked to
    if (ask) {

      # they wanted a GUI but can't support one
      if (widget & !(.Platform$OS.type == "windows" || .Platform$GUI ==
                       "AQUA" || (capabilities("tcltk") && capabilities("X11")))) {
        widget <- FALSE
      }

      k <- select.list(k, preselect=FALSE, multiple = TRUE,
                       title = "Packages to be updated", graphics = widget)

    }

    if (length(k) > 0) {

      # use the resultant list

      invisible(apply(tmp %>% filter(Package %in% k, update.available==TRUE), 1, function(p) {
        # this will prevent package install errors from stopping the function
        try(devtools::install_github(sprintf("%s/%s", as.character(p["owner"]), as.character(p["repo"]))))
      }))

      # update the package version info on post-install

      tmp$installed.version <- sapply(tmp$Package, function(x) { as.character(packageVersion(x))}, USE.NAMES=FALSE)
      tmp$update.available <- mapply(compareVersion, tmp$Version, tmp$current.version, USE.NAMES=FALSE)<0

    }

  }

  if (show.location) {
    return(tmp %>% select(package=Package, repo, owner, installed.version=Version, current.version, update.available, lib=LibPath))
  } else {
    return(tmp %>% select(package=Package, repo, owner, installed.version=Version, current.version, update.available))
  }

}


#' @title whats_new - produce a report on the new features of any out of date packages
#' @description
#'     Uses \link{render} from \link{rmarkdown} to generate a report of \code{NEWS} files from any packages that have updates not currently installed
#'
#' @import rmarkdown
#' @export
whats_new <- function() {
  browseURL(render(system.file("markdown/whatsnew.Rmd", package="dtupdate"), output_file=tempfile(fileext=".html"), quiet=TRUE))
}
